README.md
# Full Fat Monads

An experimental replacement for DRY monads that attempts to:

  - reduce implementation complexity
  - provide a more idiomatic FP usage experience
  - provide more functionality

## Change log

See [the change log](Changelog.md)

Each branch in the code is listed in the change log.  Branches that
are not merged will have a `TODO` section heading with some points
indicating what I plan to do.  Each item in the `TODO` section is
checked off as I do it.  More items are usually added over time as
I discover what I need to do.

Merged branches will have a section called `DONE` that lists all
of the things I did in that branch.  As much as possible items in the
change log will correspond to commits in git.  If you do a `git blame`
on the Changelog, you will find the commit where that work was done.
I'm not strict about that, so often it doesn't work, but I try to
do it as often as I remember.

## How to do various things

### Tests

To run the tests once
`bundle exec rspec`

To watch the directories and run the tests after changes
`bundle exec guard`

### Lint

To run the rubocop linter
`bundle exec rubocop`

To automatically fix linting warnings (warning: not always safe)
`bundle exec rubocop -A`

### Documentation

To generate the documentation
`yard doc`

To display how bad I am at writing documentation

`yard stats`


Changelog.md
# Change Log

## Branch matching
### TODO
  - [X] Review the docs for pattern matching in Ruby.
    - (Docs are here)[https://docs.ruby-lang.org/en/3.0/syntax/pattern_matching_rdoc.html]
  - [ ] Try to match `Maybe`

## Branch docs
### DONE
  - [X] Write documentation for error.rb
  - [X] Write documentation for `Escape`
  - [X] Write a slightly better README
  - [X] Write documentation for `Maybe`
    - Good enough for now

## Branch escape
### DONE
  - [X] `value!` returns the value or raises an exception which is caught by `escape`
  - [X] rename `value!` to `v!`.
    - It looks a little nicer in a void context.
    - I realised the `!` is the `not` operator, so `monad.!` is the same as `!monad` which
      could easily be confused for a binary operator.  `v!` is easy to type and doesn't look
      bad, I think.

## Branch maybe (actually in main :-P)
### DONE
  - [X] Sketch in a `Maybe` class
    - Is `Maybe` a good name?  I guess most people know what it means.
      Rust uses `Option`, which is also good.  Hmm...
  - [X] Tests
    - [X] `Mixin`
    - [X] `Some`
    - [X] `None`
  - [X] Escape
    - This is my version of "Do" notation.  It's closer to Rust's ? operator.
      Basically `monad.!` returns the value in the monad or raises an exception.
      `escape { some(monad.!) }` will catch the exception and return the monad
      that did not have the value.

## Branch setup
### DONE
  - [X] Add a Changelog
  - [X] Test that `!` is a valid identifier in Ruby
    - It's weird that `!` is fine, but `?` is not.
  - [X] Set up stuff
    - [X] ruby
      - Go with Ruby 3.2.2 for no good reason
    - [X] bundler,
    - [X] rubocop
    - [X] rspec,
    - [X] guard,
  - [X] Make a minimal gem


spec/spec_helper.rb
# frozen_string_literal: true

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See https://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
#   # This allows you to limit a spec run to individual examples or groups
#   # you care about by tagging them with `:focus` metadata. When nothing
#   # is tagged with `:focus`, all examples get run. RSpec also provides
#   # aliases for `it`, `describe`, and `context` that include `:focus`
#   # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
#   config.filter_run_when_matching :focus
#
#   # Allows RSpec to persist some state between runs in order to support
#   # the `--only-failures` and `--next-failure` CLI options. We recommend
#   # you configure your source control system to ignore this file.
#   config.example_status_persistence_file_path = "spec/examples.txt"
#
#   # Limits the available syntax to the non-monkey patched syntax that is
#   # recommended. For more details, see:
#   # https://rspec.info/features/3-12/rspec-core/configuration/zero-monkey-patching-mode/
#   config.disable_monkey_patching!
#
#   # This setting enables warnings. It's recommended, but in some cases may
#   # be too noisy due to issues in dependencies.
#   config.warnings = true
#
#   # Many RSpec users commonly either run the entire suite or an individual
#   # file, and it's useful to allow more verbose output when running an
#   # individual spec file.
#   if config.files_to_run.one?
#     # Use the documentation formatter for detailed output,
#     # unless a formatter has already been configured
#     # (e.g. via a command-line flag).
#     config.default_formatter = "doc"
#   end
#
#   # Print the 10 slowest examples and example groups at the
#   # end of the spec run, to help surface which specs are running
#   # particularly slow.
#   config.profile_examples = 10
#
#   # Run specs in random order to surface order dependencies. If you find an
#   # order dependency and want to debug it, you can fix the order by providing
#   # the seed, which is printed after each run.
#   #     --seed 1234
#   config.order = :random
#
#   # Seed global randomization in this process using the `--seed` CLI option.
#   # Setting this allows you to use `--seed` to deterministically reproduce
#   # test failures related to randomization by passing the same `--seed` value
#   # as the one that triggered the failure.
#   Kernel.srand config.seed
end


spec/ff_monads_spec.rb
# frozen_string_literal: true

require_relative '../lib/ff_monads'

RSpec.describe FFMonads do
  describe 'Mixin with extend' do
    module ModuleFoo
      extend FFMonads
    end

    it 'does something' do
      expect(ModuleFoo.some(42).to_s).to eql('some(42)')
    end
  end

  describe 'Mixin with include' do
    class ClassFoo
      include FFMonads

      def some_forty_two
        some(42)
      end
    end

    it 'does something' do
      expect(ClassFoo.new.some_forty_two.to_s).to eql('some(42)')
    end
  end
end


spec/ff_monads/maybe_spec.rb
# frozen_string_literal: true

require_relative '../../lib/ff_monads/maybe'
require_relative '../../lib/ff_monads/error'
require_relative '../../lib/ff_monads/escape'

RSpec.describe FFMonads::Maybe do
  include FFMonads::Maybe::Mixin
  extend FFMonads::Maybe::Mixin

  describe 'constructors' do
    describe 'some' do
      it 'creates a Maybe with a value' do
        expect(some(42).to_s).to eql('some(42)')
      end
    end

    describe 'none' do
      it 'creates a Maybe without a value' do
        expect(none.to_s).to eql('none()')
      end
    end
  end

  describe FFMonads::Maybe::Some do
    describe 'none?' do
      it 'always returns false' do
        expect(some(42).none?).to eql(false)
      end
    end

    describe 'some?' do
      it 'always returns true' do
        expect(some(42).some?).to eql(true)
      end
    end

    describe 'map' do
      it 'applies the value to a block, wrapping the result' do
        expect(some(42).map { |x| x + 1 }).to eql(some(43))
      end
    end

    describe 'and_then' do
      it 'applies the value to a block, returning what the block returns' do
        expect(some(42).and_then { |_| none }).to eql(none)
      end
    end

    describe 'v!' do
      it 'returns the value' do
        expect(some(42).v!).to eql(42)
      end

      context 'ignoring return value' do
        include FFMonads::Escape::Mixin

        it 'does not trigger a Rubocop error' do
          expect(
            escape do
              none.v!
              some(100)
            end
          ).to eql(none)
        end
      end
    end

    describe 'eql?' do
      context 'other is a Some containing the same value' do
        it 'returns true' do
          expect(some(42)).to eql(some(42))
        end
      end

      context 'other is a Some containing a different value' do
        it 'returns false' do
          expect(some(42)).to_not eql(some(10))
        end
      end

      context 'other is a different class' do
        it 'returns false' do
          expect(some(42)).to_not eql(42)
        end
      end
    end

    describe 'to_s' do
      it 'outputs the value as a string' do
        expect(some(42).to_s).to eql('some(42)')
      end

      it 'nests' do
        expect(some(some(42)).to_s).to eql('some(some(42))')
      end
    end
  end

  describe FFMonads::Maybe::None do
    describe 'none?' do
      it 'always returns true' do
        expect(none.none?).to eql(true)
      end
    end

    describe 'some?' do
      it 'always returns false' do
        expect(none.some?).to eql(false)
      end
    end

    describe 'map' do
      it 'returns none' do
        expect(none.map { |x| x + 1 }).to eql(none)
      end
    end

    describe 'and_then' do
      it 'returns none' do
        expect(none.and_then { |_| none }).to eql(none)
      end
    end

    describe 'v!' do
      it 'raises an exception' do
        expect { none.v! }.to raise_error(FFMonads::NoValueError)
      end
    end

    describe 'eql?' do
      context 'other is a None' do
        it 'returns true' do
          expect(none).to eql(none)
        end
      end

      context 'other is a Some containing a value' do
        it 'returns false' do
          expect(none).to_not eql(some(42))
        end
      end

      context 'other is a different class' do
        it 'returns false' do
          expect(none).to_not eql(42)
        end
      end
    end

    describe 'to_s' do
      it 'outputs the value as a string' do
        expect(none.to_s).to eql('none()')
      end
    end
  end

  describe 'pattern matching' do
    it 'matches on classes' do
      result = case some(42)
               in FFMonads::Maybe::Some
                 true
               in FFMonads::Maybe::None
                 false
               end

      expect(result).to eql(true)
    end

    it 'matches on values' do
      result = case some(42)
               in FFMonads::Maybe::Some(Integer => x)
                 x
               in FFMonads::Maybe::None
                 nil
               end

      expect(result).to eql(42)
    end

    describe 'mixing in Some and None' do
      add_classes
      it 'can use Some' do
        result = case some(42)
                 in Some(Integer => x)
                   x
                 else
                   false
                 end

        expect(result).to eql(42)
      end

      it 'can use None' do
        result = case none
                 in None
                   true
                 else
                   false
                 end

        expect(result).to eql(true)
      end
    end
  end
end


lib/ff_monads/escape.rb
# frozen_string_literal: true

require_relative 'error'

module FFMonads
  # Catch `NoValueError` exceptions, returning the monad that caused the issue.
  # @example You can mix in this module.
  #   include FFMonads::Escape::Mixin
  module Escape
    # Return early from the given block if an evaluated monad has no value.
    # @param [Proc] block The block to call.
    # @return [FFMonads::Maybe] The monad returned from the block.
    # @example Monads that can be unwrapped will simply run to the end of the block.
    #   monad = some(42)
    #   escape do
    #     value = monad.v! # monad has a value so it will continue
    #     some(value + 1)  # it will return some(43)
    #   end
    # @example Monads that can't be unwrapped will return early
    #   monad = none
    #     escape do
    #       value = monad.v! # monad does not have a value so it will return none here
    #       some(value + 1)  # this line will never be reached
    #     end
    def self.escape(&block)
      block.call
    rescue NoValueError => e
      e.source
    end

    module Mixin
      def escape(&)
        Escape.escape(&)
      end
    end
  end
end


lib/ff_monads/maybe.rb
# frozen_string_literal: true

require_relative 'error'

module FFMonads
  # Represents an optional value.
  # Maybe is normally used in conjunction with `.map` or `.and_then`
  # to operate on an optional value without crashing when the value
  # does not exist.
  # @example You can mix in this module.
  #   include FFMondas::Maybe::Mixin
  # @example A value that exists.
  #   some(42)
  # @example A value that does not exist.
  #   none
  class Maybe
    # Applies an optional value to a block.
    # It takes the value from the Maybe, transforms it,
    # puts back into a Maybe and returns it.
    # @param [Block] A block accepting the value and returning a transformed value.
    # @return [FFMonads::Maybe] The value from the block wrapped in a Maybe.
    # @example Called with some, it will always return some.
    #   some(42).map {|x| x + 1} # returns some(43)
    # @example Called with none, it will always return none.
    #   none.map {|x| x + 1} # returns none
    def map(&); end

    # Applies an optional value to a block.
    # It takes the value from the Maybe, transforms it,
    # and returns it without putting it back into a Maybe.
    # It assumes the block will put the value back into
    # a Maybe (or other monad).
    # @param [Block] A block accepting and returning a transformed value wrapped in a Monad.
    # @return [FFMonads::Maybe] Monad returned from the block.
    # @example `.and_then` allows you to choose the monad to return.
    #   # This returns none
    #   some(0).and_then do |x|
    #     if x == 0
    #       none
    #     else
    #       some(100 / x)
    #     end
    #   end
    def and_then(&); end

    # Returns true if the Maybe is none.
    # @return [Boolean]
    def none?; end

    # Returns true if the Maybe is some.
    # @return [Boolean]
    def some?; end

    # Returns the value in the Maybe.
    # Raises NoValueError if the Maybe is none.
    # @return [Something] the value in the Maybe.
    # @raise [FFMonads::NoValueError] if the Maybe is none.
    def v!; end

    # Returns true if this Maybe contains an equivalent value as the other Maybe.
    # @param [FFMonads::Maybe] other
    # @return [Boolean]
    def eql?; end

    # Returns a string representation of the Maybe.
    # @return [String]
    # @example Mostly used for #inspect.
    #   some(42).to_s # returns "some(42)"
    #   none.to_s     # returns "none"
    def to_s; end

    # Returns a hash representation of the Maybe.
    # Possibly for serialisation???
    # @return [Hash<Symbol, Value>]
    # @example
    #   some(42).to_h # returns { type: Some, value: 42 }
    #   none.to_h     # returns { type: None }
    def to_h; end

    # Returns a string representation of the Maybe
    # @see #to_s
    def inspect
      to_s
    end

    class Some < Maybe
      # rubocop: disable Lint/MissingSuper
      def initialize(value)
        @value = value
      end
      # rubocop: enable Lint/MissingSuper

      def deconstruct
        [@value]
      end

      def none?
        false
      end

      def some?
        true
      end

      def map(&block)
        self.class.new(block.call(@value))
      end

      def and_then(&block)
        block.call(@value)
      end

      def v!
        @value
      end

      def eql?(other)
        return false unless other.is_a?(Some)

        @value.eql?(other.v!)
      end

      def to_s
        "some(#{@value})"
      end

      def to_h
        { type: Some, value: @value }
      end
    end

    class None < Maybe
      def deconstruct
        []
      end

      def none?
        true
      end

      def some?
        false
      end

      def map(&)
        self
      end

      def and_then(&)
        self
      end

      def v!
        raise(NoValueError, self)
      end

      def eql?(other)
        other.is_a?(None)
      end

      def to_s
        'none()'
      end

      def to_h
        { type: None }
      end
    end

    module Mixin
      def add_classes
        pp self
        const_set('Some', Some)
        const_set('None', None)
      end

      def some(value)
        Some.new(value)
      end

      def none
        None.new
      end
    end

    include Mixin
  end
end
